## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [21]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')

df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [22]:
# View the column names in the data
df.columns


Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')

In [23]:
# Use the describe function to gather some basic statistics
df.describe()


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [24]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [25]:
# What three item categories had the most entries?
df['category'].value_counts().head(3)

category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64

In [26]:
# For the category with the most entries,
# which subcategory had the most entries?
df['subcategory'].loc[df['category']=='consumables'].value_counts().head(3)


subcategory
bathroom supplies    6424
paper products       5874
kitchen supplies     5746
Name: count, dtype: int64

In [27]:
# Which five clients had the most entries in the data?
df[['first','last', 'client_id']].value_counts().head(5)


first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [28]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = df['client_id'].value_counts().head(5)
top_5_clients

client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [29]:
# How many total units (the qty column) did the
# client with the most entries order order?
total_client = df[df['client_id']==33615]['qty'].sum()
total_client

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [30]:
# Create a column that calculates the 
# subtotal for each line using the unit_price
# and the qty

df['subtotal'] = df['unit_price']*df['qty']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  
0     1096.80     762.71         7.50  105            1  115164.00  
1       24.95      15.09         1.49   21            0     523.95  
2       13.52       7.86         1.68   39            6     527.28  
3       36.42      24.85         1.23   29            3    1056.18  
4      195.10     108.17        46.43   20            1    3902.00

In [31]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound
# for orders over 50 pounds and $10 per
# pound for items 50 pounds or under.
df['shipping_price'] = (df['qty'] * df['unit_weight']) * (((df['qty'] * df['unit_weight']) > 50)* 7 + ((df['qty'] * df['unit_weight']) <= 50) * 10)
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number   subtotal  \
0     1096.80     762.71         7.50  105            1  115164.00   
1       24.95      15.09         1.49   21            0     523.95   
2       13.52       7.86         1.68   39            6     527.28   
3       36.42      24.85         1.23   29            3    1056.18   
4      195.10     108.17        46.43   20            1    3902.00   

   shipping_price  
0         5512.50  
1          312.90  
2          458.64  
3          356.70  
4         6500.20

In [32]:
# Create a column for the total price
# using the subtotal and the shipping price
# along with a sales tax of 9.25%
df['total_price'] = (df['subtotal'] + df['shipping_price']) * 1.0925
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...     category         subcategory unit_price  \
0          17        2023  ...        decor            wall art    1096.80   
1          20        2023  ...  consumables                pens      24.95   
2           4        2023  ...     software  project management      13.52   
3          17        2023  ...  consumables                pens      36.42   
4          21        2023  ...  consumables                misc     195.10   

   unit_cost  unit_weight  qty  line_number   subtotal  shipping_price  \
0     762.71         7.50  105            1  115164.00         5512.50   
1      15.09         1.49   21            0     523.95          312.90   
2       7.86         1.68   39            6     527.28          458.64   
3      24.85         1.23   29            3    1056.18          356.70   
4     108.17        46.43   20            1    3902.00         6500.20   

     total_price  
0  131839.076250  
1     914.258625  
2    1077.117600  
3    1543.571400  
4   11364.403500  

[5 rows x 21 columns]

In [33]:
# Create a column for the cost
# of each line using unit cost, qty, and
# shipping price (assume the shipping cost
# is exactly what is charged to the client).
df['line_cost'] = df['unit_cost']*df['qty'] + df['shipping_price']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ...         subcategory unit_price unit_cost  \
0          17        2023  ...            wall art    1096.80    762.71   
1          20        2023  ...                pens      24.95     15.09   
2           4        2023  ...  project management      13.52      7.86   
3          17        2023  ...                pens      36.42     24.85   
4          21        2023  ...                misc     195.10    108.17   

   unit_weight  qty  line_number   subtotal  shipping_price    total_price  \
0         7.50  105            1  115164.00         5512.50  131839.076250   
1         1.49   21            0     523.95          312.90     914.258625   
2         1.68   39            6     527.28          458.64    1077.117600   
3         1.23   29            3    1056.18          356.70    1543.571400   
4        46.43   20            1    3902.00         6500.20   11364.403500   

   line_cost  
0   85597.05  
1     629.79  
2     765.18  
3    1077.35  
4    8663.60  

[5 rows x 22 columns]

In [34]:
# Create a column for the profit of
# each line using line cost and line price
df['profit'] = df['total_price'] - df['line_cost']
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year  ... unit_price unit_cost unit_weight  qty  \
0          17        2023  ...    1096.80    762.71        7.50  105   
1          20        2023  ...      24.95     15.09        1.49   21   
2           4        2023  ...      13.52      7.86        1.68   39   
3          17        2023  ...      36.42     24.85        1.23   29   
4          21        2023  ...     195.10    108.17       46.43   20   

   line_number   subtotal  shipping_price    total_price  line_cost  \
0            1  115164.00         5512.50  131839.076250   85597.05   
1            0     523.95          312.90     914.258625     629.79   
2            6     527.28          458.64    1077.117600     765.18   
3            3    1056.18          356.70    1543.571400    1077.35   
4            1    3902.00         6500.20   11364.403500    8663.60   

         profit  
0  46242.026250  
1    284.468625  
2    311.937600  
3    466.221400  
4   2700.803500  

[5 rows x 23 columns]

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [35]:
# Check your work using the totals above
display(df[df['order_id'] == 2742071]['total_price'].sum())
display(df[df['order_id'] == 2173913]['total_price'].sum())
display(df[df['order_id'] == 2742071]['total_price'].sum())

152811.897075

162388.7193

152811.897075

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [42]:
# How much did each of the top 5 clients by quantity
# spend? Check your work from Part 1 for client ids.
total_spending = df[df['client_id'].map(lambda x: x in top_5_clients)]
total_spending = total_spending.groupby('client_id').sum()
total_spending['total_price'].sort_values().head()

client_id
33615    8.377309e+06
46820    9.743794e+06
66037    1.025951e+07
38378    1.290655e+07
24741    8.226889e+07
Name: total_price, dtype: float64

In [43]:
# Create a summary DataFrame showing the totals for the
# for the top 5 clients with the following information:
# total units purchased, total shipping price,
# total revenue, and total profit. Sort by total profit.
summary_df = total_spending[['qty','']]


first  \
client_id                                                      
24741      KendraKendraKendraKendraKendraKendraKendraKend...   
33615      JessicaJessicaJessicaJessicaJessicaJessicaJess...   
38378      AlexandraAlexandraAlexandraAlexandraAlexandraA...   
46820      BryanBryanBryanBryanBryanBryanBryanBryanBryanB...   
66037      AngelaAngelaAngelaAngelaAngelaAngelaAngelaAnge...   

                                                        last  \
client_id                                                      
24741      GarrettGarrettGarrettGarrettGarrettGarrettGarr...   
33615      ReyesReyesReyesReyesReyesReyesReyesReyesReyesR...   
38378      YoungYoungYoungYoungYoungYoungYoungYoungYoungY...   
46820      MyersMyersMyersMyersMyersMyersMyersMyersMyersM...   
66037      EverettEverettEverettEverettEverettEverettEver...   

                                                         job  \
client_id                                                      
24741      Pension scheme managerPension scheme managerPe...   
33615      Production designer, theatre/television/filmPr...   
38378      DietitianDietitianDietitianDietitianDietitianD...   
46820      Chief of StaffChief of StaffChief of StaffChie...   
66037      OsteopathOsteopathOsteopathOsteopathOsteopathO...   

                                                       phone  \
client_id                                                      
24741      808.257.2029808.257.2029808.257.2029808.257.20...   
33615      5016915923501691592350169159235016915923501691...   
38378      265-666-3155x118265-666-3155x118265-666-3155x1...   
46820      444.233.9146x233444.233.9146x233444.233.9146x2...   
66037      729-620-6192x875729-620-6192x875729-620-6192x8...   

                                                       email    order_id  \
client_id                                                                  
24741      garrettkendra8479@jefferson.comgarrettkendra84...  1232695609   
33615      jessica.reyes@grimes.netjessica.reyes@grimes.n...  1262003546   
38378      ayoung9158@stevens.comayoung9158@stevens.comay...  1242894660   
46820      bmyers3696@wilson.combmyers3696@wilson.combmye...  1240590249   
66037      aeverett@rodriguez.comaeverett@rodriguez.comae...  1207923746   

                                                  order_date  order_week  \
client_id                                                                  
24741      2023-05-182023-05-252023-03-282023-02-052023-0...        2534   
33615      2023-05-022023-05-162023-05-242023-02-152023-0...        2612   
38378      2023-01-292023-05-072023-02-262023-01-122023-0...        2392   
46820      2023-05-122023-02-062023-02-282023-04-242023-0...        2481   
66037      2023-04-092023-03-272023-05-042023-04-092023-0...        2556   

           order_year                                            item_id  ...  \
client_id                                                                 ...   
24741          418761  ZXX97056-96-8AEMB99988-23-1MEMD25345-63-4EXMZ3...  ...   
33615          445060  BMB23158-88-9ABDC65078-60-7MZBZ76765-83-2XXBU4...  ...   
38378          418761  ZMB00381-41-4EXDM19959-88-6XEZC20385-63-6EAEA1...  ...   
46820          422807  CXU60946-73-5BEMX21429-07-6DDEA49899-87-6AMMX8...  ...   
66037          426853  XUX09951-98-0MUAB86387-64-8BUBA66080-53-5CADB8...  ...   

          unit_price unit_cost  unit_weight     qty  line_number     subtotal  \
client_id                                                                       
24741       25386.25  17900.04       936.56  239862          655  70176885.25   
33615       29872.65  21716.88      1260.94   64313          665   5839032.11   
38378       29685.22  21702.21      1189.50   73667          613   8384321.15   
46820       25212.96  18140.06      1109.48   75768          683   7317356.03   
66037       30164.95  21508.98      1001.73   43018          651   7995708.38   

           shipping_price   total_price    line_cost     

In [38]:
# Format the data and rename the columns
# to names suitable for presentation.
# Currency should be in millions of dollars.



In [39]:
# Sort the updated data by "Total Profit" form highest to lowest
